packages/firebase_database/lib/src/event.dart

```diff
@@ -17,26 +17,43 @@ enum _EventType {
 /// `Event` encapsulates a DataSnapshot and possibly also the key of its
 /// previous sibling, which can be used to order the snapshots.
 class Event {
-  Event._(this._data) : snapshot = DataSnapshot._(_data['snapshot']);
+  Event._(Map<Object, Object> _data)
+      : previousSiblingKey = _data['previousSiblingKey'],
+        snapshot =
+            DataSnapshot._fromJson(_data['snapshot'], _data['childKeys']);
 
   final DataSnapshot snapshot;
-  Map<dynamic, dynamic> _data;
 
-  String get previousSiblingKey => _data['previousSiblingKey'];
+  final String previousSiblingKey;
 }
 
 /// A DataSnapshot contains data from a Firebase Database location.
 /// Any time you read Firebase data, you receive the data as a DataSnapshot.
 class DataSnapshot {
-  DataSnapshot._(this._data);
-
-  final Map<dynamic, dynamic> _data;
+  DataSnapshot._(this.key, this.value);
+
+  factory DataSnapshot._fromJson(
+    Map<Object, Object> _data,
+    List<Object> childKeys,
+  ) {
+    Object dataValue = _data['value'];
+    Object value;
+
+    if (dataValue is Map<Object, Object>) {
+      value = {for (final key in childKeys) key: dataValue[key]};
+    } else if (dataValue is List<Object>) {
+      value = childKeys.map((key) => dataValue[int.parse(key)]).toList();
+    } else {
+      value = dataValue;
+    }
+    return DataSnapshot._(_data['key'], value);
```

Comment:
This is technically breaking behaviour. If a dev is currently using the output of `value` without the expectation it will be ordered (unlikely but possible) it could cause some issues.

I think we should create a new method on the class, called `forEach` similar to how JS works (and support breaking out of the loop too).

Instead it'd look like this:

```
class DataSnapshot {
  DataSnapshot._(this._data);

  final Map<dynamic, dynamic> _data;

  /// The key of the location that generated this DataSnapshot.
  String get key => _data['key'];

  /// Returns the contents of this data snapshot as native types.
  dynamic get value => _data['value'];

  dynamic forEach(ForEachAction action) {
    // Order data into a List & apply to action
  }
}

```

What do you think?

Reply:
IMO, I don't think it does change the behaviour. The `childKeys` returned are unordered unless you use the `orderByChild` API which will return the `childKeys` ordered. I don't mind adding `forEach` but the existing bug will still need to be rectified anyway.

Reply:
Would it make sense then to let "value" return the dynamic value as before, but add `forEach` which returns an iterator? At least then it allows users to keep existing behaviour but have a new feature which provides an actual typed Iterator?

Reply:
I may be misunderstanding, but the problem with leaving this as the same API:
```dart
 dynamic get value => _data['value'];
```

is because of the example that is illustrated in the [issue](#753):

```dart
(await FirebaseDatabase.instance
            .reference()
            .child(subpath)
            .orderByChild('year')
            .limitToFirst(3)
            .onValue
            .first)
        .snapshot
        .value
        .values
        .forEach((v) => debugPrint(v['year'].toString()));
```

This would still be unordered. I understand it's existing behaviour, but its tough to imagine a user calling `orderByChild()` and expecting an unordered list. 

On the other side, I can also appreciate that maybe somewhere out there is an existing code base  calling `orderByChild()`, perhaps accidentally, that is expecting the default query ordering. 

So I can get on board with either approach. Any thoughts @Salakar @rrousselGit?

Reply:
@Ehesp I don't think `forEach` should behave differently

It'd be very weird for `.value.forEach` to behave differently from `.forEach`

Reply:
I also do not think this is a breaking change.

Refactoring unordered to ordered is not breaking because the "unordered" order is not guaranteed. 
People shouldn't rely on the order of an unordered collection